File: frontend/src/lib/Transaction.ts
import { BigDecimal } from "./additionalTypes/BigDecimal";
import { BudgetDto } from "./Budget";
import { BudgetCategoryDto } from "./Category";

export interface Transaction {
  id?: number | null;
  budgetCategory?: BudgetCategoryDto | null;
  budget?: BudgetDto | null;
  amount?: BigDecimal | null;
  description?: string | null;
  transactionDate?: string | null;
  type?: string | null;
}

export interface TransactionRequest {
  name?: string | null;
  amount?: BigDecimal | null;
  description?: string | null;
  transactionDate?: string | null;
  type?: string | null;
  budgetCategoryId?: number | null;
}

export interface TransactionUpdateRequest extends TransactionRequest {}

File: backend/src/main/java/com/example/budgeter/entity/Transaction.java
package com.example.budgeter.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;

@Getter
@Setter
@Entity
@Table(name = "transactions")
public class Transaction {
  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "id", nullable = false)
  private Integer id;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "user_id")
  private User user;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "budget_category_id")
  private BudgetCategory budgetCategory;

  @NotNull
  @Column(name = "amount", nullable = false, precision = 12, scale = 2)
  private BigDecimal amount;

  @Column(name = "description", length = Integer.MAX_VALUE)
  private String description;

  @NotNull
  @Column(name = "transaction_date", nullable = false)
  private LocalDate transactionDate;

  @Size(max = 20)
  @NotNull
  @Column(name = "type", nullable = false, length = 20)
  private String type;


}